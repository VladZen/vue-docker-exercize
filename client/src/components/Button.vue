<template>
  <button
    class="btn"
    :type="$attrs.type || 'button'"
    :class="{
      [`btn--${variant}`]: !!variant,
    }"
    v-on="{ ...$listeners }"
  >
    <slot></slot>
  </button>
</template>

<script>
const STYLES = ["warning", "danger", "success"];

export { STYLES };
export default {
  props: {
    variant: {
      type: String,
      validator: (v) => STYLES.includes(v),
    },
  },
};
</script>

<style lang="scss">
.btn {
  background-color: transparent;
  position: relative;
  border-radius: $base-radius;
  border: $btn-border-width solid $btn-border-color;
  color: $btn-color;
  min-width: 35px;
  margin-left: $base-offset;

  &--danger {
    border-color: $danger-color;
    color: $danger-color;
  }

  &--warning {
    border-color: $warning-color;
    color: $warning-color;
  }

  &--success {
    border-color: $success-color;
    color: $success-color;
  }
}
</style>
