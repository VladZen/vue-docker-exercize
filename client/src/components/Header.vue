<template>
  <div class="layout__header">
    <div v-if="$slots.buttons" class="header__btns">
      <slot name="buttons"></slot>
    </div>

    <h1>{{ merchantName }}<br />Inventory</h1>
    <slot></slot>
  </div>
</template>

<script>
import { mapState, mapActions } from "vuex";

export default {
  created() {
    this.fetchMerchantName();
  },
  computed: {
    ...mapState(["merchantName"]),
  },
  methods: {
    ...mapActions({
      fetchMerchantName: "GET_MERCHANT",
    }),
  },
};
</script>

<style lang="scss" scoped></style>
