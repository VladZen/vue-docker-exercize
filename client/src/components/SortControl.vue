<template>
  <div class="sort-control">
    <span>Sorted by {{ sortBy.column }}:</span>

    <select v-model="sortBy" class="sort-control__select">
      <optgroup v-for="column in sortings" :key="column" :label="column">
        <option
          v-for="direction in dirs"
          :value="{ column, direction }"
          :key="direction"
        >
          {{ direction }}
        </option>
      </optgroup>
    </select>
  </div>
</template>

<script>
import { mapMutations, mapState } from "vuex";

export default {
  data() {
    return {
      sortings: ["name", "quantity"],
      dirs: ["asc", "desc"],
    };
  },
  computed: {
    ...mapState(["orderBy"]),
    sortBy: {
      get() {
        return this.orderBy;
      },
      set(value) {
        this.setOrder(value);
      },
    },
  },
  methods: {
    ...mapMutations(["setOrder"]),
  },
};
</script>

<style lang="scss" scoped>
.sort-control {
  display: flex;
  justify-content: flex-end;
  margin-bottom: $base-offset;

  &__select {
    margin-left: $base-offset;
  }
}
</style>
