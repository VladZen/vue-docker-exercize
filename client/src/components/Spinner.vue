<template>
  <div v-if="flag" class="spinner">
    <div></div>
    <div></div>
    <div></div>
  </div>
</template>

<script>
export default {
  props: {
    flag: {
      type: Boolean,
      default: false,
    },
  },
};
</script>

<style lang="scss">
.spinner {
  display: inline-block;
  position: relative;
  width: 80px;
  height: 80px;
}

.spinner div {
  display: inline-block;
  position: absolute;
  left: 8px;
  width: 16px;
  background: #fdd;
  animation: spinner 1.2s cubic-bezier(0, 0.5, 0.5, 1) infinite;
}

.spinner div:nth-child(1) {
  left: 8px;
  animation-delay: -0.24s;
}

.spinner div:nth-child(2) {
  left: 32px;
  animation-delay: -0.12s;
}

.spinner div:nth-child(3) {
  left: 56px;
  animation-delay: 0;
}

@keyframes spinner {
  0% {
    top: 8px;
    height: 64px;
  }
  50%,
  100% {
    top: 24px;
    height: 32px;
  }
}
</style>
