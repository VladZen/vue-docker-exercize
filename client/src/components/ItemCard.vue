<template>
  <div
    class="inventory-item"
    :class="{
      shake: shake,
    }"
  >
    <div
      class="inventory-item__name"
      :class="{
        invalid: $errors.name,
      }"
    >
      <input
        ref="name-input"
        type="text"
        v-model="local.name"
        :disabled="mode === 'show'"
        :placeholder="mode === 'show' ? 'No Name' : 'Type name'"
      />

      <small class="error-msg">expects not to be empty!</small>
    </div>

    <div
      v-if="mode !== 'create'"
      class="inventory-item__quantity"
      :class="{
        invalid: $errors.quantity,
      }"
    >
      <input
        type="number"
        v-model.number="local.quantity"
        :disabled="mode === 'show'"
        placeholder="Type quantity"
      />

      <small class="error-msg">expects a positive integer!</small>
    </div>

    <div class="inventory-item__controls">
      <template v-if="mode === 'create'">
        <btn
          class="btn--save"
          title="Save"
          variant="success"
          :disabled="interactionInProgress"
          @click="save"
        >
          <spinner variant="success" :flag="!loaded.create" />
          <font-awesome-icon v-if="loaded.create" icon="check" />
        </btn>

        <btn
          class="btn--edit"
          title="Discard"
          variant="danger"
          :disabled="interactionInProgress"
          @click="reset"
        >
          <font-awesome-icon icon="undo" />
        </btn>
      </template>

      <template v-else-if="mode === 'show'">
        <btn
          class="btn--edit"
          title="Edit"
          :disabled="interactionInProgress"
          @click="toggleMode('edit')"
        >
          <font-awesome-icon icon="edit" />
        </btn>

        <btn
          class="btn--remove"
          title="Remove"
          variant="danger"
          :disabled="interactionInProgress"
          @click="remove"
        >
          <spinner variant="danger" :flag="!loaded.remove" />
          <font-awesome-icon v-if="loaded.remove" icon="times" />
        </btn>
      </template>

      <template v-else-if="mode === 'edit'">
        <btn
          v-if="mode === 'edit'"
          class="btn--apply"
          title="Apply"
          variant="success"
          :disabled="interactionInProgress"
          @click="update"
        >
          <spinner variant="success" :flag="!loaded.update" />
          <font-awesome-icon v-if="loaded.update" icon="check" />
        </btn>

        <btn
          key="discard"
          class="btn--edit"
          title="Discard"
          variant="danger"
          :disabled="interactionInProgress"
          @click="
            toggleMode('show');
            reset(item);
          "
        >
          <font-awesome-icon icon="undo" />
        </btn>
      </template>
    </div>
  </div>
</template>

<script>
import _isEqual from "lodash.isequal";
import defaultItem from "@/config/default-item.js";

import { reactive, ref, computed, watch } from "@vue/composition-api";
import useFlag from "@/use/requestLoadingFlag";
import useStore from "@/use/store";

import Spinner from "@/components/Spinner";
import Btn from "@/components/Button";

export default {
  components: {
    Spinner,
    Btn,
  },
  setup(props, ctx) {
    const store = useStore(ctx);
    const mode = ref(props.state);
    const local = ref(defaultItem(props.item));
    const $errors = reactive({ name: false, quantity: false });
    const shake = ref(false);

    const { loaded: create, promise: dispatchCreate } = useFlag(
      store.dispatch.bind(null, "ADD")
    );
    const { loaded: update, promise: dispatchUpdate } = useFlag(
      store.dispatch.bind(null, "UPDATE")
    );
    const { loaded: remove, promise: dispatchRemove } = useFlag(
      store.dispatch.bind(null, "REMOVE")
    );
    const loaded = reactive({
      create,
      update,
      remove,
    });
    const interactionInProgress = computed(() =>
      Object.keys(loaded).some((action) => !loaded[action])
    );

    watch(
      () => props.item,
      (value) => (local.value = { ...value })
    );

    return {
      loaded,
      dispatchCreate,
      dispatchUpdate,
      dispatchRemove,
      interactionInProgress,
      defaultItem: () => reactive(defaultItem(props.item)),
      local,
      mode,
      $errors,
      shake,
    };
  },
  props: {
    item: {
      type: Object,
      default: () => {},
    },
    state: {
      type: String,
      validator: (value) => ["show", "create", "edit"].includes(value),
      default: "show",
    },
  },
  methods: {
    async save() {
      if (this.isInvalid()) return;
      const id = await this.dispatchCreate(this.local.name);
      this.$emit("save", id);
      this.reset();
    },
    async remove() {
      await this.dispatchRemove(this.local);
      this.$emit("remove", this.local.id);
    },
    async update() {
      if (this.isInvalid()) return;

      if (!_isEqual(this.item, this.local)) {
        this.local = await this.dispatchUpdate(this.local);
        this.$emit("update", this.local.id);
      }
      this.toggleMode("show", this.local.id);
    },
    toggleMode(mode, id = this.local.id) {
      this.mode = mode;
      this.$emit("update:mode", { mode, id });
    },
    reset(to = null) {
      this.local = this.defaultItem(to);
      this.$errors = {
        name: false,
        quantity: false,
      };
      this.$emit("reset");
    },
    shakeItem() {
      this.shake = true;
      setTimeout(() => (this.shake = false), 1000);
    },
    isInvalid() {
      this.$errors.name = !this.local.name;
      if (this.mode === "edit")
        this.$errors.quantity =
          ["", null, undefined].includes(this.local.quantity) ||
          this.local.quantity < 0;

      const error = this.$errors.name || this.$errors.quantity;
      if (error) this.shakeItem();
      return error;
    },
  },
};
</script>

<style lang="scss">
.inventory-item {
  display: flex;
  align-items: flex-end;
  justify-content: space-between;
  background-color: #fff;
  border-radius: $base-radius;
  padding: 5px $base-offset * 2;
  min-height: 83px;

  &__quantity {
    margin-left: 10px;
    display: inline-flex;
    align-items: baseline;
    margin-right: $base-offset;

    input {
      text-align: right;
    }
  }

  &__quantity,
  &__name {
    display: flex;
    flex-direction: column;
  }

  &__name {
    flex: 2;
  }

  &__quantity {
    flex: 1;
  }

  &__controls {
    align-self: center;
  }

  &.shake {
    animation: shake 1s cubic-bezier(0.36, 0.07, 0.19, 0.97) both;
    transform: translate3d(0, 0, 0);
    backface-visibility: hidden;
    perspective: 1000px;
  }

  .error-msg {
    opacity: 0;
    color: $danger-color;
    transition: opacity $transition;
    pointer-events: none;
  }

  .invalid {
    input {
      border-color: $danger-color;
    }

    .error-msg {
      opacity: 1;
      pointer-events: all;
    }
  }

  & + & {
    margin-top: $base-offset;
  }
}

@keyframes shake {
  10%,
  90% {
    transform: translate3d(-1px, 0, 0);
  }

  20%,
  80% {
    transform: translate3d(2px, 0, 0);
  }

  30%,
  50%,
  70% {
    transform: translate3d(-4px, 0, 0);
  }

  40%,
  60% {
    transform: translate3d(4px, 0, 0);
  }
}
</style>
