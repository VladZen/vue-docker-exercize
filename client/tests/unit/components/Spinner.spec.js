import { mount } from "@vue/test-utils";
import Spinner from "@/components/Spinner";
import { STYLES } from "@/components/Spinner";

const mountSpinner = (variant, flag = false) => {
  return mount(Spinner, {
    propsData: {
      variant: variant || undefined,
      flag,
    },
  });
};

describe("Spinner", () => {
  it('is hidden by default', () => {
    expect(mountSpinner(undefined, false)).toMatchSnapshot();
  });

  it.each(STYLES)("checks basic rendering of variant %s", (variant) => {
    expect(mountSpinner(variant, true)).toMatchSnapshot();
  });
});
