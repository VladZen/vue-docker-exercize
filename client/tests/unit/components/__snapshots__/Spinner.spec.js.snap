// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner checks basic rendering of variant danger 1`] = `
<div class="spinner spinner--danger">
  <div class="bounce1"></div>
  <div class="bounce2"></div>
  <div class="bounce3"></div>
</div>
`;

exports[`Spinner checks basic rendering of variant success 1`] = `
<div class="spinner spinner--success">
  <div class="bounce1"></div>
  <div class="bounce2"></div>
  <div class="bounce3"></div>
</div>
`;

exports[`Spinner checks basic rendering of variant warning 1`] = `
<div class="spinner spinner--warning">
  <div class="bounce1"></div>
  <div class="bounce2"></div>
  <div class="bounce3"></div>
</div>
`;

exports[`Spinner is hidden by default 1`] = ``;
